File: src/MinecraftHelper.cpp
#include <iostream>

#include "MinecraftHelper.h"
#include "_internal.h"

using nlohmann::json;

static const std::string minecraft_manifest = "https://launchermeta.mojang.com/mc/game/version_manifest.json";

namespace mc {
    using namespace _internal;

    std::optional<std::string> getLatestVersion(bool verbose) {
        std::optional<std::string> result = get_url(minecraft_manifest, verbose);
        if (!result.has_value()) return result;
        else return std::optional(json::parse(result.value())["latest"]["release"]);
    }

    std::optional<json> getVersion(const std::string& version_number, bool verbose) {
        std::optional<std::string> urlResult = get_url(minecraft_manifest, verbose);

        if (!urlResult.has_value()) return urlResult;

        json j = json::parse(urlResult.value());

        if (!j["versions"].is_array()) {
            if (verbose) std::cerr << "versions is not an array!" << std::endl;
            return std::nullopt;
        }

        auto v = j["versions"].get<std::vector<json>>();
        for (json j2 : v) {
            if (j2["id"] == version_number) return j2;
        }
        
        return std::nullopt;
    }

}

File: Sources/fabricversions/MinecraftHelper.swift
import Foundation

fileprivate let MINECRAFT_MANIFEST = "https://launchermeta.mojang.com/mc/game/version_manifest.json"
fileprivate let target = URL(string: MINECRAFT_MANIFEST)!

public var VERSION_MANIFEST: VersionManifest? = nil

public struct VersionManifest : Decodable {
  public struct LatestVersions : Decodable {
    let release: String
  }

  public struct Version : Decodable {
    public let id: String
    public let url: String
  }

  let latest: LatestVersions
  let versions: [Version]
}

public struct MinecraftVersion : Decodable {
  let assets: String
}

fileprivate func getVersionManifest() -> Result<VersionManifest, Error> {
  if VERSION_MANIFEST == nil {
    switch sendRequest(target: target) {
      case .failure(let error): return .failure(error)
      case .success(let data): do {
        VERSION_MANIFEST = try decoder.decode(VersionManifest.self, from: data)
      } catch {
        return .failure(error)
      }
    }
  }

  return .success(VERSION_MANIFEST!)
}

public func getLatestVersion() -> Result<String, Error> {
  switch getVersionManifest() {
    case Result.failure(let error): return .failure(error)
    case Result.success(let manifest): return .success(manifest.latest.release)
  }
}

public func getVersion(number: String) -> Result<MinecraftVersion, Error> {
  switch getVersionManifest() {
    case .failure(let error): return .failure(error)
    case .success(let manifest):
      if let first = (manifest.versions.first { $0.id == number }) {
        switch sendRequest(target: URL(string: first.url)!) {
          case .failure(let error): return .failure(error)
          case .success(let data): do {
            return .success(try decoder.decode(MinecraftVersion.self, from: data))
          } catch {
            return .failure(error)
          }
        }
      } else {
        return .failure(VersionError.runtimeError("Cound not find Minecraft version \(number)!"))
      }
  }
}
